crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -244,3 +244,145 @@ impl OutputNoteTracker {
         self.output_notes.into_iter().flatten().collect()
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::{
+        mock_proven_tx,
+        note::{mock_note, mock_output_note, mock_unauthenticated_note_commitment},
+    };
+
+    #[test]
+    fn test_output_note_tracker_duplicate_output_notes() {
+        let mut txs = mock_proven_txs();
+
+        OutputNoteTracker::new(&txs).expect("Something went wrong, no error expected");
+
+        let duplicate_output_note = txs[1].output_notes().get_note(1).clone();
+
+        txs.push(mock_proven_tx(
+            3,
+            vec![],
+            vec![duplicate_output_note.clone(), mock_output_note(8), mock_output_note(4)],
+        ));
+
+        match OutputNoteTracker::new(&txs) {
+            Err(BuildBatchError::DuplicateOutputNote(note_id, _)) => {
+                assert_eq!(note_id, duplicate_output_note.id())
+            },
+            res => panic!("Unexpected result: {res:?}"),
+        }
```

Comment:
Pity that `assert_matches!` is not yet stable; but maybe we should use the [assert_matches crate](https://docs.rs/assert_matches/latest/assert_matches/) until it is. 

Its a common occurrence I think? If not then these once off matches are fine.

Reply:
Nice catch, but I don't think this worth of using one more crate. Let's rewrite it once `assert_matches!` is stabilized.